### PR TITLE
Fix MSVC Y2K38 bug in from_time_t

### LINKED ISSUE #87: from_time_t is unnecessarily forcing 32-bit seconds
See: https://github.com/boostorg/date_time/pull/35#issuecomment-353632702
https://github.com/boostorg/date_time/commit/b1dc2b306ea6abb645b0e334e3d5bb1b5ed80818

### PR DESCRIPTION
**Pull request by orudge**
from_time_t was altered recently to cast the incoming value to a long. On LLP64 platforms, such as Windows, this truncates the value to 32-bit. My proposed patch casts to int64_t.

I've also added a unit test involving a 64-bit time_t conversion.

This fixes #87

### CODE REVIEW

**Comment by mclow** [comment]
I'll take a look. Note that your tests are failing on some platforms.  I *suspect* that casting to long was the wrong approach as well - but I've got to do some investigation first.

Also - your test addition probably needs to be inside an `#ifdef` block, rather than `if (...)` because it might throw warnings (or even errors) on systems w/o 64 bit ints.

**Comment by orudge** [comment]
Thanks, I'll check what's going on with the unit tests and will look at putting it into an #ifdef too.

**Comment by orudge** [comment]
I've updated the code to fix the #ifdef issue and ensure the time_t variable is definitely being populated with a 64-bit value, but it seems this is still failing - the internal seconds value in the ptime still seems to be truncated to 32-bit. I don't really understand why to be honest; I'm able to step through using MSVC 2017 (x86 build) and it's all working OK for me. This fix was also sufficient to sort out our application.

It may be there's something deeper down in the headers causing issues, but unfortunately I'm a bit out of my depth trying to solve that. If my code can be of use then please free free to adapt it. At the very least the time_t to long truncation definitely appears to be wrong, but whether my fix is the 'right' one I'm not sure.

**Comment by jeking3** [comment]
Internally the code converts to nanoseconds, and 64 bits is insufficient to hold the conversion.  The issue is down in `boost/date_time/time_resolution_traits.hpp`, specifically in to_tick_count.  The unit test case added causes an int64_t overflow.

**Comment by mclow** [comment]
2^64 nanoseconds is ~584 years. Are we really testing for +/- 600 years?

**Comment by jeking3** [comment]
Yes.  I added a test for the max date to time_t conversation and it fails.
Looking into what I can do without getting too invasive.

**Comment by jeking3** [comment]
@mclow current unit tests only test conversion to and from time_t up to year 2032 because the library only supported 32-bit time_t (when it was created).  I added a test locally to do time_t conversion to/from the maximum supported date (year 9999) and it fails because internally ticks are nanoseconds, and there isn't enough space for the conversions.  I'm experimenting with putting in a 128 bit internal ticks right now, but we'll see.. that may not be the best answer.

**Comment by ValZapod** [comment]
Can we do this now and add ~584 years support later?

**Comment by jeking3** [comment]
@orudge rebase on develop for a complete CI scan

File: include/boost/date_time/posix_time/conversion.hpp [hunk 1]
```diff
@@ -26,7 +26,7 @@ namespace posix_time {
   inline
   ptime from_time_t(std::time_t t)
   {
-    return ptime(gregorian::date(1970,1,1)) + seconds(static_cast<long>(t));
+    return ptime(gregorian::date(1970,1,1)) + seconds(static_cast<boost::int64_t>(t));
```
Comment (jeking3) [inline include/boost/date_time/posix_time/conversion.hpp:29]: See alternative suggestion in https://github.com/boostorg/date_time/issues/87 - is the cast necessary here?

**Review by jeking3 — APPROVED** [review]
If you look back through the history of this line, it was originally without a seconds cast.  Then it was cast to long to avoid a compiler error, which likely happened right before a change was made in 1.67.0 that made the cast unnecessary, so we're back to not having a cast again.
 https://github.com/boostorg/date_time/blob/acbee0dc6a168289c398041e91e053b3684083fc/include/boost/date_time/time_resolution_traits.hpp#L86-L94

**Comment by jeking3** [comment]
It's likely worth combing through some of the builds to see the state of warnings around that change before merging.

**Pull request merged**